### PR TITLE
Implement accounts

### LINKED ISSUE #11: Implement accounts

### PR DESCRIPTION
**Pull request by niklaslong**
Closes #11.

### CODE REVIEW

**Comment by niklaslong** [comment]
@fancycade This is getting pretty close I think. 

I've started thinking about further abstractions for improving the API: `Olm` module, empty other than the `version` function; `Account`, `Session`, `Utility` etc... This PR just contains `Olm` and `Account` and the latter is just a transparent wrapper for now. In future I'd like to decode the Jason responses and perhaps allow functions to be piped. 

The C code is fairly close, though I'm getting a memory leak when unpickling an account first with a bad key, then with the correct one. To reproduce:

```elixir
iex(3)> {:ok, a} = Account.create()
{:ok, #Reference<0.2513799406.1192361985.17566>}
iex(4)> {:ok, p} = Account.pickle(a, "key")
{:ok,
 "0sfvEHVhTvhjAQaYbj76HIoKcnB4nL0HXWvC7YdmgjSXLRTJ7QmaFivhN0MITo1ukTYXZnC9ALgXlgACVYj7Vfw5X5kxV/ui2Fhrg4lK6ry2FAWDDB7JhytWjYUFwxiPe5xzic/o4vEpDxbMAAGwDNpniiRpdMJJ28VENnPeeUxvD7qu0HhYPAf0cPuZ9v3j6Q6wfdidkVb9KkVrR7ww+eCNClgd80LMD5kziJCCjKzVHLmoRc9HHQ"}
iex(5)> Account.unpickle(p, "wrong_key")
{:error, 'BAD_ACCOUNT_KEY'}
iex(6)> Account.unpickle(p, "key")
{:error, 'BAD_ACCOUNT_KEY'}
iex(7)> p
<<68, 106, 255, 185, 163, 113, 127, 255, 255, 159, 252, 255, 255, 250, 233, 207,
  255, 255, 254, 64, 102, 253, 255, 248, 255, 255, 200, 159, 255, 255, 249, 174,
  194, 252, 255, 231, 255, 191, 223, 175, 255, 252, 255, 255, 255, 95, 255, 212,
  255, 255, ...>>
```

It seems like C is trying to decode the pickled account passed in and mutating the memory location. I'm guessing copying data in memory to a new buffer would solve that. I'm inclined to leave that as a follow up PR considering how large this one is.

File: c_src/olm_nif.c [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+    char   bytes[random_length];
+
+    size_t result = olm_create_account(account, bytes, random_length);
+
+    // Return {:ok, account_ref} or {:error, last_error}.
+    if (result == olm_error()) {
+        ERL_NIF_TERM error_atom    = enif_make_atom(env, "error");
+        ERL_NIF_TERM error_message = enif_make_string(
+            env, olm_account_last_error(account), ERL_NIF_LATIN1);
+
```
Comment (fancycade) [inline c_src/olm_nif.c:71]: I think you want to release the account resource here as well.

File: c_src/olm_nif.c [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+        return enif_make_tuple2(env, error_atom, error_message);
+    } else {
```
Comment (fancycade) [inline c_src/olm_nif.c]: I would remove the else. The if will short circuit the expression with the return statement anyways.

File: c_src/olm_nif.c [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+
+    size_t result = olm_pickle_account(
+        account, key.data, key.size, pickled.data, pickled_length);
+
+    // Return {:ok, pickled} or {:error, last_error}.
+    if (result == olm_error()) {
+        ERL_NIF_TERM error_atom    = enif_make_atom(env, "error");
+        ERL_NIF_TERM error_message = enif_make_string(
+            env, olm_account_last_error(account), ERL_NIF_LATIN1);
+
```
Comment (fancycade) [inline c_src/olm_nif.c:108]: Need enif_release_binary here I believe.

File: c_src/olm_nif.c [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+        ERL_NIF_TERM ok_atom = enif_make_atom(env, "ok");
+        ERL_NIF_TERM term    = enif_make_resource(env, account);
+        enif_release_resource(account);
+
+        return enif_make_tuple2(env, ok_atom, term);
+    }
+}
+
+static ERL_NIF_TERM
+account_idenitiy_keys(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
```
Comment (fancycade) [inline c_src/olm_nif.c]: Misspelled account_identity_keys

File: c_src/olm_nif.c [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+
+    size_t result = olm_account_identity_keys(
+        account, identity_keys.data, identity_keys.size);
+
+    // Returns {:ok, identity_keys} or {:error, last_error}.
+    if (result == olm_error()) {
+        ERL_NIF_TERM error_atom    = enif_make_atom(env, "error");
+        ERL_NIF_TERM error_message = enif_make_string(
+            env, olm_account_last_error(account), ERL_NIF_LATIN1);
+
```
Comment (fancycade) [inline c_src/olm_nif.c:181]: call release_binary

File: c_src/olm_nif.c [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+    size_t result = olm_unpickle_account(
+        account, key.data, key.size, pickled.data, pickled.size);
+
+    // Return {:ok, account_ref} or {:error, last_error}.
+    if (result == olm_error()) {
+        ERL_NIF_TERM error_atom    = enif_make_atom(env, "error");
+        ERL_NIF_TERM error_message = enif_make_string(
+            env, olm_account_last_error(account), ERL_NIF_LATIN1);
+
```
Comment (fancycade) [inline c_src/olm_nif.c:146]: call release_resource

**Review by fancycade — COMMENTED** [review]
Not sure where the memory leak might be cropping up. But I did notice a few things.

I commented on a few places, but the overall pattern applies to most of these functions. 
I noticed you aren't freeing the resources/binary on the error case, so that will definitely cause a memory leak. The alloc_binary either requires a make or a release call. 

Also, not super necessary but I would remove the else expressions, since you have a return in the if's.

File: c_src/olm_nif.c (hunk 4 above)
Comment (niklaslong) [reply c_src/olm_nif.c]: Good catch!

**Comment by niklaslong** [comment]
I've released resources and binaries where appropriate. I've also fixed (I think) the leak issue for `unpickle` by copying the pickled account into a local variable and passing that to `olm_unpickle_account` instead of the pointer to the data in erlang/elixir: 

https://github.com/niklaslong/olm-elixir/blob/a3e61082165392213cf9dd2657d5a8be8bed318f/c_src/olm_nif.c#L126-L129

**Review by fancycade — APPROVED** [review]
:+1: Looks good to me!

**Pull request merged**